core/signature_receiver.go

```diff
@@ -0,0 +1,329 @@
+package core
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+	"slices"
+	"time"
+
+	"github.com/Layr-Labs/eigensdk-go/logging"
+)
+
+// signatureReceiver is a struct for receiving SigningMessages for a single batch. It should never be instantiated
+// manually: it exists only as a helper struct for the ReceiveSignatures method.
+type signatureReceiver struct {
+	logger               logging.Logger
+	indexedOperatorState *IndexedOperatorState
+
+	signerMap                 map[OperatorID]bool
+	aggregateSignatures       map[QuorumID]*Signature
+	aggregateSignersG2PubKeys map[QuorumID]*G2Point
+
+	stakeSigned map[QuorumID]*big.Int
+
+	batchHeaderHash    [32]byte
+	signingMessageChan chan SigningMessage
+	quorumIDs          []QuorumID
+}
+
+// ReceiveSignatures receives SigningMessages over the signingMessageChan, and yields QuorumAttestations produced
+// from these SigningMessages.
+//
+// The yielded QuorumAttestations contain aggregate signing data from all SigningMessages received thus far. Each
+// QuorumAttestation will have incorporated more SigningMessages than the previously yielded QuorumAttestation.
+//
+// The caller should not manually close the returned QuorumAttestation chan. It will be closed automatically, when:
+// 1. The global attestation timeout is exceeded
+// 2. A SigningMessage from every Operator has been received and processed
+//
+// Before being closed, the QuorumAttestation chan will have returned a QuorumAttestation containing data from every
+// gathered SigningMessage.
+func ReceiveSignatures(
+	ctx context.Context,
+	logger logging.Logger,
+	indexedOperatorState *IndexedOperatorState,
+	batchHeaderHash [32]byte,
+	signingMessageChan chan SigningMessage,
+) (chan *QuorumAttestation, error) {
+	sortedQuorumIDs, err := getSortedQuorumIDs(indexedOperatorState)
+	if err != nil {
+		return nil, fmt.Errorf("get sorted quorum ids: %w", err)
+	}
+
+	signerMap := make(map[OperatorID]bool)
+	aggregateSignatures := make(map[QuorumID]*Signature, len(sortedQuorumIDs))
+	aggregateSignersG2PubKeys := make(map[QuorumID]*G2Point, len(sortedQuorumIDs))
+
+	// initialized stakeSigned map with 0 stake signed for each quorum
+	stakeSigned := make(map[QuorumID]*big.Int, len(sortedQuorumIDs))
+	for _, quorumID := range sortedQuorumIDs {
+		stakeSigned[quorumID] = big.NewInt(0)
+	}
+
+	receiver := &signatureReceiver{
+		logger:                    logger,
+		indexedOperatorState:      indexedOperatorState,
+		aggregateSignatures:       aggregateSignatures,
+		signerMap:                 signerMap,
+		aggregateSignersG2PubKeys: aggregateSignersG2PubKeys,
+		stakeSigned:               stakeSigned,
+		batchHeaderHash:           batchHeaderHash,
+		signingMessageChan:        signingMessageChan,
+		quorumIDs:                 sortedQuorumIDs,
+	}
+
+	attestationChan := make(chan *QuorumAttestation)
```

Comment:
This creates a channel with capacity 0 (i.e. the goroutine that inserts a thing into the channel blocks until another goroutine removes the thing from the channel). If we know the maximum number of messages that will be passed, we can avoid the unnecessary blocking by setting the capacity equal to the maximum number of messages.

```suggestion
	attestationChan := make(chan *QuorumAttestation, len(signerMap))
```

Reply:
done